### PR TITLE
[perf] Route slow (non-fast) tokenizers (e.g. Kimi) through `.encode()` on the `--enable-dynamic-batch-tokenizer` path (completes #25265)

### LINKED ISSUE #29163: [Perf] Slow (non-fast) tokenizers still use the slow HF `__call__` path under `--enable-dynamic-batch-tokenizer`
### Summary

A non-fast (`is_fast=False`) tokenizer that ships a fast `encode()` — e.g. Kimi's `TikTokenTokenizer` (a `trust_remote_code` slow `PreTrainedTokenizer` backed by tiktoken/Rust) — is ~6–7x slower than necessary on the encode path when `--enable-dynamic-batch-tokenizer` is **on**. [#25265](https://github.com/sgl-project/sglang/pull/25265) already routes such tokenizers through `.encode()` in `TokenizerManager._tokenize_texts`, but **not on the dynamic-batch path**: there the request goes through `AsyncDynamicbatchTokenizer`, which invokes the tokenizer via `__call__` → the stock `_encode_plus` → `convert_tokens_to_ids(tokenize(text))` slow chain, even though `encode()` already returns the final `list[int]` straight from tiktoken (Rust).

This work runs on the `TokenizerManager` event loop path, so the redundant cost sits directly on the per-request TTFT critical path. Removing it is a straight latency win — the same motivation as #25265, which already fixed the non-dynamic-batch branch.

### Root cause / where the gap is

`TokenizerManager._tokenize_texts`:
- When `--enable-dynamic-batch-tokenizer` is **off**, #25265 routes slow (non-`is_fast`) tokenizers through `self.tokenizer.encode(...)` (fast). ✅
- When it is **on**, single-string requests are routed to `AsyncDynamicbatchTokenizer.encode()`, which internally calls `self.tokenizer(prompts, ...)` / `self.tokenizer(p, **kw)` — i.e. `__call__`, not `.encode()` → the slow `_encode_plus`/`tokenize` chain. #25265's fast branch is never reached. ❌

```
async_dynamic_batch_tokenizer.py:
    encode_fn = partial(self.tokenizer, prompts, **kwargs)   # __call__  -> slow
    self.tokenizer(p, **kw) for p, kw in zip(...)            # __call__  -> slow
```

Also note: `AsyncDynamicbatchTokenizer` is constructed with **no `is_fast` gate** (only on the flag / `skip_tokenizer_init`), so slow tokenizers get wrapped and routed through `__call__`.

### Confirmed end-to-end

Instantiating the real upstream `AsyncDynamicbatchTokenizer` with the real Kimi tokenizer and awaiting `.encode(...)`, with a counter on the slow `tokenize()`:
- a **single** request triggers `tokenize()` **once**;
- **4 concurrent** requests coalesce into **one** `tokenizer([4])` batch call that triggers `tokenize()` **4 times**.

i.e. the dynamic-batch path goes through the slow chain in practice (measured on transformers 5.x, CPU — no model weights / GPU).

### Still open upstream

No PR/issue addresses this path. #25265 (regular path) and #25953 (chat-template encode site) are the only related encode-side fixes, and neither covers the dynamic-batch path; `AsyncDynamicbatchTokenizer` itself has never been changed to handle slow tokenizers.

### Why the slow path is pure overhead for this tokenizer

Stock `_encode_plus` builds ids via `convert_tokens_to_ids(tokenize(text))`. For Kimi's tokenizer, `tokenize()` runs a Python trie split over all special tokens and then round-trips every token `int -> str -> int` across a ~160k vocab, while `encode()` already returns the final ids from tiktoken with `allowed_special="all"`.

### Measurements (real Kimi-K2.5 tokenizer)

byte-for-byte identical output; slow `tokenize()` is skipped entirely on the `.encode()` fast path.

| input tokens | stock `__call__` | `.encode()` fast path | speedup |
|---:|---:|---:|---:|
| 8,000 | 21.3 ms | 2.9 ms | 7.3x |
| 60,000 | 160.5 ms | 22.0 ms | 7.3x |
| 120,000 | 316.0 ms | 43.8 ms | 7.2x |
| batch ×16 | 2.13 ms | 0.62 ms | 3.5x |

Consistent with #25265 (80k: 308→47 ms, ~6.5x) and the equivalent vLLM report (vllm-project/vllm#39590, ~2.4x).

### Proposed fix

Make `AsyncDynamicbatchTokenizer` route non-`is_fast` tokenizers through `self.tokenizer.encode()` (per item, inside its existing executor), **mirroring #25265** and preserving the wrapper's off-loop behavior. Fast tokenizers keep the batched `__call__` (Rust `encode_batch`). Because the wrapper only handles single-string, non-cross-encoder inputs, there is no `token_type_ids` to worry about.

(An alternative — overriding `_encode_plus` for the model class — also works, but it monkeypatches a `trust_remote_code` class and its batch coverage depends on transformers-version-specific dispatch; the call-site `.encode()` routing is consistent with the upstream precedent and version-robust.)

PR: <link to the draft PR>

### Related

- #25265 / #25953 — encode-side fix for the regular and chat-template paths; **this issue is the remaining dynamic-batch path** (upstream has been fixing slow-tokenizer encode sites one by one).
- #16427 — detokenizer-side speedup (`SGLANG_PATCH_TOKENIZER` flag).
- #9382 — Dynamic Batch Tokenizer (the affected wrapper).
- vllm-project/vllm#39590 — same issue in vLLM.
- **Model-side context**: newer Kimi model versions ship a fast tokenizer (so `is_fast=True` and this path is naturally fine), but currently-deployed versions (e.g. Kimi-K2.5 / Kimi-K2-Instruct) still use the slow tokenizer, and that migration has had downstream breakage (exo-explore/exo#2085, earendil-works/pi#5159; HF `moonshotai/Kimi-K2.5` discussion #7 unresolved) — so a framework-side fix still matters for deployed models.

### PR DESCRIPTION
**Pull request by CZYCW**
## Motivation

[#25265](https://github.com/sgl-project/sglang/pull/25265) made slow tokenizers fast in `TokenizerManager._tokenize_texts` by routing non-`is_fast` tokenizers through `tokenizer.encode()` instead of the slow `__call__` — but **only on the branch taken when `--enable-dynamic-batch-tokenizer` is off**. With dynamic batching on, single-string requests go through `AsyncDynamicbatchTokenizer`, which invokes the tokenizer via `__call__`, so #25265's fast path is bypassed.

For any non-fast (`is_fast=False`) tokenizer that ships a fast `encode()` — e.g. Kimi's `TikTokenTokenizer`, or a `trust_remote_code` slow `PreTrainedTokenizer` backed by tiktoken (Rust):

- `__call__` runs the stock `_encode_plus` → `convert_tokens_to_ids(tokenize(text))` chain — a Python trie split plus an `int → str → int` round-trip over a ~160k vocab;
- `encode()` already returns the final `list[int]` straight from Rust.

That redundant work sits on the per-request TTFT critical path. See #29163.

This is **not Kimi-specific** — it completes #25265's pattern (*slow tokenizer → use `.encode()`*) on the one path it didn't reach:

- #25265 fixed the regular path;
- #25953 fixed the chat-template encode site;
- the dynamic-batch path is the remaining one.

## What this PR does

Two changes — a behavior fix, and a no-behavior-change refactor that the fix motivated.

### 1. Fix — route slow tokenizers through `.encode()` on the dynamic-batch path

- For non-`is_fast` tokenizers, build ids **per item via `self.tokenizer.encode(p)`** instead of `__call__`.
- The work still runs inside the wrapper's existing single-thread `ThreadPoolExecutor`, so tokenization stays off the event loop — **the wrapper's offload is preserved**.
- Returns the same dict shape the caller expects (`{"input_ids": ...}`).
- Fast tokenizers keep the existing batched `__call__` (the Rust `tokenizers` `encode_batch` / rayon path) **unchanged**.
- The dynamic batch tokenizer only handles `SINGLE_STRING`, non-cross-encoder inputs (`tokenizer_manager.py:747-750`), so `token_type_ids` is not involved.

### 2. Refactor — tidy `_process_dynamic_batch` (no behavior change)

Adding the slow-tokenizer branch made an existing smell acute: the method mixed *deciding how to encode* (`is_fast` / `can_batch`) with *running it and resolving futures*, and repeated the `if not fut.done(): fut.set_result(...)` guard three times. It now reads top-down:

- `_build_encode_fn(prompts, kwargs_list, can_batch)` — picks the strategy **once** (slow `.encode()` / uniform-kwargs batched `__call__` / per-item `__call__`) and returns a zero-arg callable producing one result dict per prompt.
- `_set_results` / `_set_exception` — centralize the `fut.done()` cancellation guard and the batch-level exception fan-out.
- The batched `__call__` split preserves **all** keys the tokenizer returns (`input_ids`, `attention_mask`, `token_type_ids`, …), not just `input_ids`.

## Correctness & performance

Correctness rests on the same `tokenizer.encode()` that #25265's regular-path fix uses, byte-for-byte verified against the stock slow chain (short/long inputs, `add_special_tokens` True/False, truncation, batch lists, edge cases — including empty/whitespace/CJK/emoji and special-token literals; `.encode()` and `__call__` produce identical ids on all of them).

The win is the same one #25265 reports, now extended to the dynamic-batch path: routing a slow Kimi tokenizer through tiktoken's native `.encode()` instead of the `__call__` → `_encode_plus` → `convert_tokens_to_ids(tokenize(text))` chain. `.encode()` fast path vs stock `__call__` on the real `moonshotai/Kimi-K2.5` tokenizer:

| input tokens | stock `__call__` | `.encode()` fast path | speedup |
|---:|---:|---:|---:|
| 8,000 | 21.3 ms | 2.9 ms | 7.3x |
| 60,000 | 160.5 ms | 22.0 ms | 7.3x |
| 120,000 | 316.0 ms | 43.8 ms | 7.2x |
| batch ×16 | 2.13 ms | 0.62 ms | 3.5x |

Consistent with #25265 (80k: 308→47 ms, ~6.5x) and the equivalent vLLM report (vllm-project/vllm#39590). Fast tokenizers keep the batched `__call__` unchanged.

> Validated end-to-end through the real `AsyncDynamicbatchTokenizer` with the real Kimi tokenizer (see Tests): output is byte-for-byte equal to the stock `__call__` **and** to #25265's `.encode()`, and the slow `tokenize()` is never called. The ms numbers above isolate the underlying `.encode()`-vs-`__call__` cost (the call this PR swaps) and do not include the small async-wrapper overhead. The refactor itself is cost-neutral.

## Tests

**Unit — CI** (`test/registered/unit/managers/test_async_dynamic_batch_tokenizer_slow.py`, 12 cases, CPU, no model download). A minimal char-level slow tokenizer whose `encode()` returns ids directly stands in for tiktoken.

- Dispatch:
  - non-fast tokenizer (single & concurrently-batched) → ids byte-for-byte equal to the stock `__call__`, slow `tokenize()` never called;
  - missing `is_fast` attribute → treated as slow;
  - fast tokenizer → stays on batched `__call__`, and the per-prompt split preserves every key (`attention_mask`, `token_type_ids`, not just `input_ids`);
  - non-empty kwargs → fall back to `__call__`;
  - heterogeneous kwargs → run per item and emit the batching-disabled warning;
  - single fast request with kwargs → runs per item, no warning.
- Resolution:
  - a future cancelled by a disconnected client is skipped while siblings still resolve;
  - a tokenizer error fans out to every pending future;
  - a pre-cancelled future is left cancelled when the batch errors.

Run: `python -m pytest test/registered/unit/managers/test_async_dynamic_batch_tokenizer_slow.py`

**Manual** (`test/manual/test_async_dynamic_batch_tokenizer.py`, 14 cases) also passes. Its mock tokenizer was a bare function with no `is_fast`/`encode`, which the new slow branch would route to a nonexistent `.encode()`; it is now marked `is_fast=True` (a fast-tokenizer stand-in) so it stays on the `__call__` path it was written to exercise.

Closes #29163

### CODE REVIEW

**Comment by ronhuafeng** [comment]
I checked this PR against the current PR head (`7e6e76068`) and the local result looks good.

Validation I ran:

```text
python3 -m py_compile python/sglang/srt/managers/async_dynamic_batch_tokenizer.py test/registered/unit/managers/test_async_dynamic_batch_tokenizer_slow.py test/manual/test_async_dynamic_batch_tokenizer.py
git diff --check HEAD^..HEAD
PYTHONPATH=python python3 -m pytest test/registered/unit/managers/test_async_dynamic_batch_tokenizer_slow.py test/manual/test_async_dynamic_batch_tokenizer.py -q
```

Result:

```text
26 passed, 4 warnings in 25.88s
```

One small environment note: the SGLang image I used had `pytest` but not `pytest-asyncio`, so the manual async tests need that plugin available. Without it, pytest reports async-test collection failures before exercising those tests; with the plugin installed, both the registered unit test and manual tokenizer test pass.

I also checked the current red GitHub checks. The representative `pr-gate` failure is gate-only: the PR is ready (`PR Draft: false`) but has no labels (`PR Labels: []`), so the workflow stops at `Missing required label 'run-ci'`. Lint is already green.

So from my side this looks like it is waiting on maintainer labeling/review rather than a code/test failure.
